Second iteration of Admin API auth: client authoriser

m1kola opened the pull request:
Not ready for review.

Includes https://github.com/Azure/ARO-RP/pull/158 which needs to be merged first.

jim-minter commented on pkg/util/clientauthorizer/admin.go:
```diff
@@ -5,23 +5,51 @@ package clientauthorizer
-		log: log,
+func NewAdmin() (ClientAuthorizer, error) {
+	roots := x509.NewCertPool()
+	ok := roots.AppendCertsFromPEM([]byte(os.Getenv("ADMIN_CA_CERTS")))
```
I'd like to get away from this being an environment variable, but maybe this will do for now.  Let's discuss quickly.

jim-minter commented on pkg/util/clientauthorizer/admin.go:
```diff
@@ -5,23 +5,51 @@ package clientauthorizer
-func NewAdmin(log *logrus.Entry) ClientAuthorizer {
-	return &admin{
-		log: log,
+func NewAdmin() (ClientAuthorizer, error) {
```
We should check ADMIN_CA_CERTS and ADMIN_CLIENT_CERT_COMMON_NAME (again) in NewAdmin so we are sure it fails closed if they are invalid (e.g. empty) + unit test.

jim-minter commented on pkg/util/clientauthorizer/admin.go:
```diff
@@ -5,23 +5,51 @@ package clientauthorizer
+	}, nil
 }
 
 func (a *admin) IsAuthorized(cs *tls.ConnectionState) bool {
```
Needs in-package unit test coverage before merge.

m1kola commented:
@jim-minter please for now ignore the changes in the `deploy` package and the way we work with the certs\env variables.

I've been learning how we deploy the RP into production and my plan is to read the CA bundle from key vault in the `cse` and mount it into the container (will be something similar to `mdm.pem` and `mdsd.pem`).

jim-minter commented:
> I've been learning how we deploy the RP into production and my plan is to read the CA bundle from key vault in the cse and mount it into the container (will be something similar to mdm.pem and mdsd.pem).

If you're doing this, better to follow the pattern in #228 that reads from KV in the RP binary rather than the CSE.

*But* I don't think this is applicable here - we want a non-secret list of root certs, and the KV provides a single cert chain or a secret limited to around 25KB.

nilsanderselde commented on pkg/util/clientauthorizer/admin.go:
```diff
@@ -5,23 +5,68 @@ package clientauthorizer
+
+	_, err := cs.PeerCertificates[0].Verify(verifyOpts)
+	if err != nil {
+		return false
```
Would we ever want to know what this error is?

m1kola replied:
Yeah, I thought about it too. I'm not sure how useful this error is: it will mean that the certificate is not valid. If someone, somehow will be bombarding us with an invalid client certificate - we will have a lot of noise in the logs. On the other after the expiery of valid certificate we will be declining requests to the admin API without logging it.

m1kola replied:
I think, I'll add logging here. The situation when someone will be sending invalid certificate on porpuse is very unlikely because admin API is not exposed to public, but it is useful to have something in logs in case we miss certificate expiration date or something else goes wrong on our side.

Thank you for raising this concern, Nils!

m1kola commented on pkg/deploy/generator/rp.go:
```diff
@@ -512,14 +531,17 @@ ExecStart=/usr/bin/docker run \
   rp
 ExecStop=/usr/bin/docker stop -t 3600 %N
-TimeoutStopSec="3600"
+TimeoutStopSec=3600
```
Current, quoted value is being ignored:

> /etc/systemd/system/aro-rp.service:10: Failed to parse TimeoutStopSec= parameter, ignoring: "3600"

m1kola commented:
@mjudeikis @jim-minter I rebased on top of master to include #204 and updated `pkg/deploy`. I tested my changes by deployed RP in production mode in our subscription, but could you please take a look to verify that I'm not doing something silly in this package?

I'll be adding tests for `pkg/util/clientauthorizer/admin.go` mow.

m1kola commented:
@mjudeikis @nilsanderselde @jim-minter I updated the PR and I think it is finally ready for review.

I tested the prod deploymnet with a self-generated certificate and everything seems to work as expected. I also updated the secrets to contain a self-signed cert we can use in shared dev environments (+ updated the doc).

Please take a look.

m1kola commented:
Something got merged, so I had to rebase to fix conflicts.

jim-minter commented on deploy/rp-development.json:
```diff
@@ -2,6 +2,12 @@
+        },
+        "adminApiClientCertCommonName": {
+            "type": "string"
+        },
```
We don't honour the admin API authentication settings in dev, therefore I don't think rp-development.json should change, or prepare-a-shared-rp-development-environment.md, or deploy-shared-env.sh?

m1kola replied:
Right! We don't even deploy vmss in shered env.

jim-minter commented on pkg/deploy/generator/rp.go:
```diff
@@ -448,6 +470,8 @@ MDM_ACCOUNT=AzureRedHatOpenShiftRP
 MDM_NAMESPACE=BBM
 CLUSTER_MDM_ACCOUNT=AzureRedHatOpenShiftCluster
 CLUSTER_MDM_NAMESPACE=BBM
+ADMIN_API_CA_BUNDLE_PATH='$ADMIN_API_CA_BUNDLE_PATH'
```
the aro-monitor service shouldn't need these - don't add them.

jim-minter commented on pkg/deploy/generator/rp.go:
```diff
@@ -346,6 +357,12 @@ chmod 0600 /etc/mdsd.pem
 az logout
 
+mkdir /etc/rp
+ADMIN_API_CA_BUNDLE_PATH=/etc/rp/admin-ca-bundle.pem
```
I am ok with hard-coding this in the codebase rather than adding another env var.  Please can we standardise on /etc/aro-rp ?

jim-minter commented on pkg/deploy/generator/rp.go:
```diff
@@ -346,6 +357,12 @@ chmod 0600 /etc/mdsd.pem
+ADMIN_API_CA_BUNDLE_PATH=/etc/rp/admin-ca-bundle.pem
+base64 -d <<<"$ADMINAPICABUNDLE" >$ADMIN_API_CA_BUNDLE_PATH
+chown -R 1000:1000 /etc/rp
+chmod 0600 $ADMIN_API_CA_BUNDLE_PATH
```
chmod not needed - it's not secret.

jim-minter commented on pkg/env/env.go:
```diff
@@ -52,19 +52,30 @@ func NewEnv(ctx context.Context, log *logrus.Entry) (Interface, error) {
 		"MDM_ACCOUNT",
 		"MDM_NAMESPACE",
+		"ADMIN_API_CA_BUNDLE_PATH",
+		"ADMIN_API_CLIENT_CERT_COMMON_NAME",
```
Put this in in cmd/aro/rp.go wrapped by `if _, ok := _env.(env.Dev); !ok` so it doesn't trigger for aro-monitor.  We already take this approach in cmd/aro/monitor.go.  It's disgusting but we'll fix it out later.

jim-minter commented on pkg/env/env.go:
```diff
@@ -52,19 +52,30 @@ func NewEnv(ctx context.Context, log *logrus.Entry) (Interface, error) {
 	}
 
-	return newProd(ctx, log, im, clientauthorizer.NewARM(log), clientauthorizer.NewAdmin(log))
+	adminClientAuthorizer, err := clientauthorizer.NewAdmin(
```
oh, I see, this really is forcing a tricky issue.  I think we need to separate this a bit so that when aro-monitor calls NewEnv in prod, it doesn't go through all of this code path.  There's actually more items in this category - currently aro-monitor pulls secrets which it never needs at startup because aro-rp does, and it shouldn't do that either.  Don't need to fix the whole issue in this PR, but try not to add a pattern which wouldn't work for the secrets problem as well.

Perhaps stop taking the clientauthorizers as input to the constructor and instead initializing them on demand in {Arm,Admin}ClientAuthorizer() instead?  Consider sync/Once.

jim-minter commented on pkg/util/clientauthorizer/admin_test.go:
```diff
@@ -0,0 +1,161 @@
+	if err != nil {
+		t.Fatal(err)
+	}
+	invalidKeyUsagesCert[0].ExtKeyUsage = []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth}
```
set the isClient argument to GenerateKeyAndCertificate above to false?

jim-minter commented on pkg/util/clientauthorizer/admin_test.go:
```diff
@@ -0,0 +1,161 @@
+	if err != nil {
+		t.Fatal(err)
+	}
+	invalidExpiredCert[0].NotAfter = time.Now().Add(-time.Hour)
```
I don't think you can really do this - I'd expect the certificate not to validate the signature it contains will now be invalid.  I think by here we're sort of testing Go library functionality rather than our functionality, so I think this case can be dropped.

m1kola replied:
This works. We will get: "x509: certificate has expired or is not yet valid".

But I agree: it's not necessary because it is something that go standard library provides for us and should have test coverage for.

jim-minter commented on pkg/util/clientauthorizer/admin_test.go:
```diff
@@ -0,0 +1,161 @@
+			wantCommonName: "validclient",
+			want:           true,
+			cs: &tls.ConnectionState{
+				PeerCertificates: append(validCertWithIntermediates, validIntermediateCaCerts...),
```
the append may have side-effects - explicitly do `validCertWithIntermediates = append(validCertWithIntermediates, validIntermediateCaCerts...)` on line 46 instead

jim-minter commented on pkg/util/clientauthorizer/admin_test.go:
```diff
@@ -0,0 +1,161 @@
+
+	for _, tt := range []struct {
+		name           string
+		wantCommonName string
```
remove `want`.  Better, can move this into the body of the test as it's constant over all test cases.

jim-minter commented on pkg/util/clientauthorizer/admin_test.go:
```diff
@@ -0,0 +1,161 @@
+	for _, tt := range []struct {
+		name           string
+		wantCommonName string
+		cs             *tls.ConnectionState
```
comment: a `func() (*tls.ConnectionState, error)` would arguably have clarified here, but no need to change what's in the PR

m1kola replied:
@jim-minter Not sure that I understand what you mean here.

jim-minter commented on pkg/util/clientauthorizer/admin_test.go:
```diff
@@ -0,0 +1,161 @@
+	utiltls "github.com/Azure/ARO-RP/pkg/util/tls"
+)
+
+func TestAdminClientAuthorizer(t *testing.T) {
```
This is excellent work!

mjudeikis commented:
/azp run e2e

m1kola commented on pkg/util/clientauthorizer/admin.go:
```diff
@@ -5,25 +5,88 @@ package clientauthorizer
 
 func (a *admin) IsReady() bool {
-	return true
+	return a.roots != nil
```
There should be no data race, right? Or do I miss something?

m1kola commented on pkg/util/clientauthorizer/admin.go:
```diff
@@ -5,25 +5,88 @@ package clientauthorizer
+func (a *admin) readCABundle(caBundlePath string) {
+	caBundle, err := a.readFile(caBundlePath)
+	if err != nil {
+		a.log.Error(err)
```
If we can't read or decode ca cert - we log an error and return. The authorizer will be not ready witohut ca cert and declining the requests.

I'm not super happy about the fact that we do not return an error while creating the authorizer now (so we can fail on the startup). I didn't manage to come up with a good way of failing on the startup with the new way of instantiating `env`. An alternative would be to exit on the first call to `env.AdminClientAuthorizer()`, but I think it's a bad idea too: will make failures less obvious.

Any thoughts?

jim-minter replied:
I think that my suggestion to use sync.Once was a bad one because it hides the error.

Let's add another function to env called `InitializeAuthorizers() error` and call that early in cmd/aro/rp.go .  It calls NewAdmin which now returns (ClientAuthorizer, error).  It also calls NewARM but it's OK that this doesn't return any error right now.  You get the validation error early.  Anyone who tries to use an authorizer without having called the initialize will dereference a null pointer and die.  IsReady() can go back to returning true and the tests simplify again.

m1kola commented on pkg/deploy/generator/resources.go:
```diff
@@ -740,8 +757,10 @@ ExecStart=/usr/bin/docker run \
+  -e ADMIN_API_CLIENT_CERT_COMMON_NAME \
   -e RP_MODE \
   -p 443:8443 \
+  -v ${ADMIN_API_CA_BUNDLE_PATH}:${ADMIN_API_CA_BUNDLE_PATH} \
```
Let's mount only what we actually need for now. It will be easier to mount more later than the other way around, I think.

jim-minter commented on pkg/util/clientauthorizer/admin.go:
```diff
@@ -5,25 +5,91 @@ package clientauthorizer
+// This authorizer allows connections only if they
+// contain a valid client certificate signed by `caBundlePath` and
+// the client certificate's CommonName equals `clientCertCommonName`.
+func NewAdmin(log *logrus.Entry, caBundlePath, clientCertCommonName string) (ClientAuthorizer, error) {
```
this doesn't matter, ignore
